src/components/validation/validators/component_spec/sources.rs

```diff
@@ -0,0 +1,392 @@
+use std::fmt::{Display, Formatter};
+
+use bytes::BytesMut;
+use vector_core::event::{Event, MetricKind};
+use vector_core::EstimatedJsonEncodedSizeOf;
+
+use crate::components::validation::{encode_test_event, TestEvent, ValidationConfiguration};
+
+use crate::components::validation::runner::config::TEST_SOURCE_NAME;
+
+use super::filter_events_by_metric_and_component;
+
+pub enum SourceMetrics {
+    EventsReceived,
+    EventsReceivedBytes,
+    ReceivedBytesTotal,
+    SentEventsTotal,
+    SentEventBytesTotal,
+}
+
+impl SourceMetrics {
+    const fn name(&self) -> &'static str {
+        match self {
+            SourceMetrics::EventsReceived => "component_received_events_total",
+            SourceMetrics::EventsReceivedBytes => "component_received_event_bytes_total",
+            SourceMetrics::ReceivedBytesTotal => "component_received_bytes_total",
+            SourceMetrics::SentEventsTotal => "component_sent_events_total",
+            SourceMetrics::SentEventBytesTotal => "component_sent_event_bytes_total",
+        }
+    }
+
+    pub const fn validation(
+        &self,
+    ) -> fn(
+        &ValidationConfiguration,
+        &[TestEvent],
+        &[Event],
+        &[Event],
+    ) -> Result<Vec<String>, Vec<String>> {
+        match self {
+            SourceMetrics::EventsReceived => validate_component_received_events_total,
+            SourceMetrics::EventsReceivedBytes => validate_component_received_event_bytes_total,
+            SourceMetrics::ReceivedBytesTotal => validate_component_received_bytes_total,
+            SourceMetrics::SentEventsTotal => validate_component_sent_events_total,
+            SourceMetrics::SentEventBytesTotal => validate_component_sent_event_bytes_total,
+        }
+    }
+
+    pub const fn all() -> [SourceMetrics; 5] {
+        [
+            SourceMetrics::EventsReceived,
+            SourceMetrics::EventsReceivedBytes,
+            SourceMetrics::ReceivedBytesTotal,
+            SourceMetrics::SentEventsTotal,
+            SourceMetrics::SentEventBytesTotal,
+        ]
+    }
+}
+
+impl Display for SourceMetrics {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.name())
+    }
+}
+
+fn validate_component_received_events_total(
+    _configuration: &ValidationConfiguration,
+    inputs: &[TestEvent],
+    _outputs: &[Event],
+    telemetry_events: &[Event],
+) -> Result<Vec<String>, Vec<String>> {
+    let mut errs: Vec<String> = Vec::new();
+
+    let metrics = filter_events_by_metric_and_component(
+        telemetry_events,
+        SourceMetrics::EventsReceived,
+        TEST_SOURCE_NAME,
+    )?;
+
+    let mut events: f64 = 0.0;
```

Comment:
Maybe I'm missing something, but how could we have a not whole number of events? Perhaps there's a better name.